A50-xds-outlier-detection.md

```diff
@@ -0,0 +1,207 @@
+
+The subchannel wrapper will track the latest state update from the underlying subchannel. By default, it will simply pass those updates along. The wrapper will have the following methods:
+
+ - `eject()`: The wrapper will report a state update with the `TRANSIENT_FAILURE` state, and will stop passing along updates from the underlying subchannel.
+ - `uneject()`: The wrapper will report a state update with the latest update from the underlying subchannel, and resume passing along updates from the underlying subchannel.
+
+### Success Rate Algorithm
+
+ 1. If the number of addresses with request volume of at least `success_rate_request_volume` is less than `success_rate_minimum_hosts`, stop.
+ 2. Calculate the mean and standard deviation of the fractions of successful requests among addresses with total request volume of at least `success_rate_request_volume`.
+ 3. For each address:
+    1. If the percentage of ejected addresses is greater than `max_ejection_percent`, stop.
+    2. If the address's total request volume is less than `success_rate_request_volume`, continue to the next address.
+    3. If the address's success rate is less than `(mean - stdev * (success_rate_stdev_factor / 1000))`, then choose a random integer in `[0, 100)`. If that number is less than `enforcing_success_rate`, eject that address.
+
+### Failure Percentage Algorithm
+
+ 1. If the number of addresses is less than `failure_percentage_minimum_hosts`, stop.
+ 2. For each address:
+    1. If the percentage of ejected addresses is greater than `max_ejection_percent`, stop.
+    2. If the address's total request volume is less than `failure_percentage_request_volume`, continue to the next address.
+    3. If the address's failure percentage is greater than `failure_percentage_threshold`, then choose a random integer in `[0, 100)`. If that number is less than `enforcing_failure_percentage`, eject that address.
+
+### xDS Integration
+
+If the `Cluster` message for an underlying cluster has the `outlier_detection` field set, the `xds_cluster_resolver` will generate an `outlier_detection` LB policy config from it for each priority as the top-level LB policy for that priority. In that case, the child policy config will be the config that would otherwise be at the top level for the priority. When `outlier_detection` is configured, the LB policy tree will look like this when using round robin, and the `outlier_detection` LB policy will be inserted similarly when using ring hash:
+
+![gRPC Client Architecture Diagram](A50_graphics/grpc_xds_client_architecture.png)
+
+[Link to SVG file](A50_graphics/grpc_xds_client_architecture.svg)
+
+### Temporary environment variable protection
+
+During initial development, the xDS part of this feature will be enabled by the `GRPC_XDS_EXPERIMENTAL_ENABLE_OUTLIER_DETECTION` environment variable. This
```

Comment:
I think we'll want the same env var to guard even the non-xDS part of this feature, so let's remove the `XDS_` from the env var name.